### PR TITLE
Swapped html table with plotly Heatmap with mouse-over event

### PR DESCRIPTION
**Pull request by lukaa12**
KNTK-230

### CODE REVIEW

File: presentation/matrix_view.py [hunk 1]
```diff
@@ -4,72 +4,124 @@
...
-def make_data(mesh: MeshResults, matrix: Matrix) -> List[Dict]:
-    data = []
-    for row in mesh.rows:
-        row_data = {"agent_alias": row.alias}
+def make_data(mesh: MeshResults, matrix: Matrix, latency_deteriorated_ms: int, latency_failed_ms: int) -> List[Dict]:
+    x = matrix.agents
+    y = [i for i in reversed(matrix.agents)]
+    z = []
+
+    for row in reversed(mesh.rows):
```
Comment (Fenthick) [inline presentation/matrix_view.py:41]: Please extract function, e.g. `make_z()` returning the numpy z array

File: presentation/matrix_view.py [hunk 2]
```diff
@@ -4,72 +4,124 @@
...
+            color = get_color(
+                matrix.cells[row.alias][col.alias].latency_microsec.value, latency_deteriorated_ms, latency_failed_ms
+            )
+            z_col.append(color)
+        z.append(z_col)
+    for i in range(len(mesh.rows)):
+        z[-(i + 1)].insert(i, None)
+    z = np.array(z)
+
+    colorscale = get_colorscale(z)
```
Comment (Fenthick) [inline presentation/matrix_view.py:53]: IMO this variable is redundant

File: presentation/matrix_view.py [hunk 3]
```diff
@@ -4,72 +4,124 @@
...
         for col in row.columns:
-            latency_ms = matrix.cells[row.alias][col.alias].latency_microsec.value / 1000  # scale to milliseconds
-            row_data[col.alias] = f"{latency_ms} ms"
-        data.append(row_data)
+            color = get_color(
+                matrix.cells[row.alias][col.alias].latency_microsec.value, latency_deteriorated_ms, latency_failed_ms
+            )
+            z_col.append(color)
+        z.append(z_col)
+    for i in range(len(mesh.rows)):
```
Comment (Fenthick) [inline presentation/matrix_view.py:49]: Please add a comment or even better extract those two lines to a function called like `unset_diagonal_color()`

File: presentation/matrix_view.py [hunk 4]
```diff
@@ -4,72 +4,124 @@
...
-                        "backgroundColor": "green",
-                    }
+def make_annotations(mesh: MeshResults, matrix: Matrix) -> List[Dict]:
+    annotations = []
+    for row in reversed(mesh.rows):
+        for col in row.columns:
+            annotations.append(
+                dict(
+                    showarrow=False,
+                    text=f"<b>{(matrix.cells[row.alias][col.alias].latency_microsec.value * 1e-3):.2f} ms<b>",
```
Comment (Fenthick) [inline presentation/matrix_view.py:80]:
```suggestion
                    text=f"<b>{(matrix.cells[row.alias][col.alias].latency_microsec.value * 1e-3):.2f} ms</b>",
```
I guess this is an HTML tag

File: requirements.txt [hunk 5]
```diff
@@ -3,4 +3,5 @@ mypy>=0.790
 python_dateutil >= 2.5.3
 setuptools >= 21.0.0
 urllib3 >= 1.25.3
-toml>=0.10.2
\ No newline at end of file
+toml>=0.10.2
+pandas >= 1.2.4
```
Comment (Fenthick) [inline requirements.txt:7]: Shouldn't it be numpy? I don't see pandas usage

**Review by Fenthick — CHANGES_REQUESTED** [review]

File: requirements.txt (hunk 5 above)
Comment (mmac-m3a) [reply requirements.txt:7]: Both numpy and pandas are heavy dependencies. It is better to avoid them unless they are used to provide functionality that would be hard to get otherwise.

File: presentation/matrix_view.py [hunk 6]
```diff
@@ -4,72 +4,124 @@
...
-            row_data[col.alias] = f"{latency_ms} ms"
-        data.append(row_data)
+            color = get_color(
+                matrix.cells[row.alias][col.alias].latency_microsec.value, latency_deteriorated_ms, latency_failed_ms
+            )
+            z_col.append(color)
+        z.append(z_col)
+    for i in range(len(mesh.rows)):
+        z[-(i + 1)].insert(i, None)
+    z = np.array(z)
```
Comment (mmac-m3a) [inline presentation/matrix_view.py:51]: AFAICT, `Plotly.Graph`should work with normal iterables (https://dash.plotly.com/dash-core-components/graph). If you confirm that `np.array` is really not required, reimplement `get_colorscale` using built in `all` function and remove dependency on numpy. `get_colorscale` may look like this:

```
def get_colorscale(z: List) ->; List[List]:
    if all([_x == 0.0 for _x in z]):
        return [[0.0, GREEN], [1.0, GREEN]]
    if all([_x == 0.5 for _x in z]):
        return [[0.0, ORANGE], [1.0, ORANGE]]
    if all([_x == 1.0 for _x in z]):
        return [[0.0, RED], [1.0, RED]]
    if all([_x != 0.0 for _x in z]):
        return [[0.0, ORANGE], [1.0, RED]]
    if all([_x != 1.0 for _x in z]):
        return [[0.0, GREEN], [1.0, ORANGE]]
    else:
        return [[0.0, GREEN], [0.5, ORANGE], [1.0, RED]]
```

**Review by mmac-m3a — COMMENTED** [review]
Remove dependency on numpy if possible.

File: presentation/matrix_view.py [hunk 7]
```diff
@@ -4,72 +4,124 @@
...
+
+def make_hover_text(mesh: MeshResults, matrix: Matrix) -> List[List[str]]:
+    text = []
+    for row in reversed(mesh.rows):
+        text_col = []
+        for col in row.columns:
+            latency_ms = matrix.cells[row.alias][col.alias].latency_microsec.value * 1e-3
+            jitter = matrix.cells[row.alias][col.alias].jitter.value
+            loss = matrix.cells[row.alias][col.alias].packet_loss.value
+            text_col.append(f"Latency: {latency_ms:.2f} ms, \nJiter: {jitter}, \nLoss: {loss}")
```
[the remark this answers is not in the document]
Comment (mateuszmidor) [reply presentation/matrix_view.py:98]: Jitter comes as microseconds, please scale to milliseconds and add "ms" suffix. 
Loss is in %, please add "%" suffix.
This is HTML, <br> tag should get you a newline

File: requirements.txt (hunk 5 above)
Comment (lukaa12) [reply requirements.txt:7]: The code used to use pandas while I was implementing this, but now it do not use it. Neither numpy nor pandas are required now. This line will be removed.

File: presentation/matrix_view.py (hunk 6 above)
Comment (lukaa12) [reply presentation/matrix_view.py:51]: Yes, numpy is not required now, I reimplemented `get_colorscale`

File: presentation/matrix_view.py (hunk 4 above)
Comment (lukaa12) [reply presentation/matrix_view.py:80]: Yes you are right. I fixed this typo

**Review by mmac-m3a — APPROVED** [review]

**Comment by mmac-m3a** [comment]
Can you, please, rebase the pull request based on the top of master?

**Comment by lukaa12** [comment]
@mmac-m3a I rebased PR to top of the master

File: infrastructure/config/config_ini.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    def latency_deteriorated_ms(self) -> int:
+        return self._latency_deteriorated_ms
+
+    @property
+    def latency_failed_ms(self) -> int:
+        return self._latency_failed_ms
+
+    @property
+    def data_update_period_seconds(self) -> int:
+        return self._data_update_period_seconds
```
Comment (mateuszmidor) [inline infrastructure/config/config_ini.py:30]: config_ini.py is already removed

File: infrastructure/config/config_toml.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    def latency_failed_ms(self) -> int:
+        return self._latency_failed_ms
+
+    @property
+    def data_update_period_seconds(self) -> int:
+        return self._data_update_period_seconds
+
+    @property
+    def data_update_lookback_minutes(self) -> int:
+        return self._data_update_lookback_minutes
```
Comment (mateuszmidor) [inline infrastructure/config/config_toml.py:35]: config_toml.py is already removed

File: requirements.txt [hunk 10]
```diff
@@ -3,4 +3,6 @@ mypy>=0.790
 python_dateutil >= 2.5.3
 setuptools >= 21.0.0
 urllib3 >= 1.25.3
-PyYAML>=5.4.1
\ No newline at end of file
+PyYAML>=5.4.1
+toml>=0.10.2
```
Comment (mateuszmidor) [inline requirements.txt:7]: toml no longer used; please remove

File: presentation/matrix_view.py (hunk 1 above)
Comment (Fenthick) [reply presentation/matrix_view.py]: Done

File: presentation/matrix_view.py (hunk 2 above)
Comment (Fenthick) [reply presentation/matrix_view.py]: Done

**Review by mateuszmidor — APPROVED** [review]

File: presentation/matrix_view.py (hunk 4 above)
Comment (Fenthick) [reply presentation/matrix_view.py]: Done

File: requirements.txt (hunk 5 above)
Comment (Fenthick) [reply requirements.txt]: Done

**Review by Fenthick — APPROVED** [review]

**Comment by Fenthick** [comment]
https://github.com/kentik/sla_dashboard_webapp/pull/10#discussion_r633753579 left

File: presentation/matrix_view.py (hunk 3 above)
Comment (lukaa12) [reply presentation/matrix_view.py]: I added it, I think it was deleted during resolving conflicts by accident

**Review by Fenthick — APPROVED** [review]

File: presentation/matrix_view.py [hunk 11]
```diff
@@ -1,80 +1,127 @@
-from typing import Dict, List
+import itertools
```
Comment (mmac-m3a) [inline presentation/matrix_view.py:1]: (non-critical) Since you need only the `chain` function, I suggest to do `from itertools import chain`

File: requirements.txt [hunk 12]
```diff
@@ -3,4 +3,4 @@ mypy>=0.790
 python_dateutil >= 2.5.3
 setuptools >= 21.0.0
 urllib3 >= 1.25.3
-PyYAML>=5.4.1
```
Comment (mmac-m3a) [inline requirements.txt:6]: I wonder what has changed here ...

**Review by mmac-m3a — APPROVED** [review]
